commons/zenoh-util/src/lib_search_dirs.rs

```diff
@@ -0,0 +1,230 @@
+//
+// Copyright (c) 2024 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+use std::{env, error::Error, fmt::Display, path::PathBuf, str::FromStr};
+
+use serde::{
+    de::{value::MapAccessDeserializer, Visitor},
+    Deserialize, Serialize,
+};
+
+use crate::lib_loader::LIB_DEFAULT_SEARCH_PATHS;
+
+#[derive(Clone, Debug, Serialize, Deserialize, Eq, Hash, PartialEq)]
+#[serde(default)]
+pub struct LibSearchDirs(Vec<LibSearchDir>);
+
+impl LibSearchDirs {
+    pub fn from_paths<T: AsRef<str>>(paths: &[T]) -> Self {
+        Self(
+            paths
+                .iter()
+                .map(|s| LibSearchDir::Path(s.as_ref().to_string()))
+                .collect(),
+        )
+    }
+
+    pub fn from_specs<T: AsRef<str>>(paths: &[T]) -> Result<Self, serde_json::Error> {
+        let dirs = paths
+            .iter()
+            .map(|s| {
+                let de = &mut serde_json::Deserializer::from_str(s.as_ref());
+                LibSearchDir::deserialize(de)
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        Ok(Self(dirs))
+    }
+}
+
+#[derive(Debug)]
+pub struct InvalidLibSearchDir {
+    found: LibSearchDir,
+    source: String,
+}
+
+impl Display for InvalidLibSearchDir {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "invalid library search directory `{:?}`: {}",
+            self.found, self.source
+        )
+    }
+}
+
+impl Error for InvalidLibSearchDir {}
+
+pub struct IntoIter {
+    iter: std::vec::IntoIter<LibSearchDir>,
+}
+
+impl Iterator for IntoIter {
+    type Item = Result<PathBuf, InvalidLibSearchDir>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next().map(LibSearchDir::into_path)
+    }
+}
+
+impl IntoIterator for LibSearchDirs {
+    type Item = Result<PathBuf, InvalidLibSearchDir>;
+
+    type IntoIter = IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        IntoIter {
+            iter: self.0.into_iter(),
+        }
+    }
+}
+
+impl Default for LibSearchDirs {
+    fn default() -> Self {
+        let de = &mut serde_json::Deserializer::from_str(&LIB_DEFAULT_SEARCH_PATHS);
```

Comment:
Here we are assuming `JSON` format which we shouldn't.

Reply:
Why shouldn't we assume JSON? If not JSON then something else will be assumed. The preview colon-separated list was also an assumption inspired by path separators in Unix (Windows uses semicolons).

Before it was a colon-separated list of strings. I changed it to JSON to re-use the `Deserialize` impl and to support the new `{ "kind": "...", "value": "..." }` syntax. We can switch it to YAML or TOML.

If it weren't for the fact that this is a compile-time configuration option, I could simply do away with this deserialization and use a `const`.

Reply:
I do not like to use `serde_json` specific things as it imposes a format.

The config per-se is a DSL, the representation must not matter so I would avoid any format-specific methods.

Reply:
> I do not like to use `serde_json` specific things as it imposes a format.
> 
> The config per-se is a DSL, the representation must not matter so I would avoid any format-specific methods.

`LIB_DEFAULT_SEARCH_PATHS` is a `const` `String` computed at compile time from an environment variable (see `zconfigurable`). It has nothing to do with Zenoh configuration.

Reply:
Here's the relevant snippet:

https://github.com/ZettaScaleLabs/zenoh/blob/f7eed49016c112c59304338d01a183a58bd5fca2/commons/zenoh-util/src/lib_loader.rs#L28-L42

Reply:
Nvm, I'm wrong, what I meant is that this environment variable HAS to be in JSON format, but do we really need it?

Can't just be a const like you mentioned in the first comment?

Reply:
> Nvm, I'm wrong, what I meant is that this environment variable HAS to be in JSON format, but do we really need it?
> 
> Can't just be a const like you mentioned in the first comment?

Yes, it can be `const` if no one needs to override it at build time...

Reply:
That's what I'm questioning it there a use-case when it is needed to override?